### ISSUE TITLE
Core-owned local model adaptation pipeline

### ISSUE THREAD

**Issue by rkfsociety** [rkfsociety/EvoHime#156]
## Цель

Добавить в EvoHime supervised Core-owned pipeline для подготовки производных локальных моделей под конкретную Windows-машину: конвертация формата, квантизация, проверка полученного артефакта и обязательное сравнение с исходной моделью через существующие hardware-fit/calibration/benchmark контуры.

MVP должен закрывать наиболее практичный путь `source model -> compatible local format -> quantized variant -> verification -> benchmark -> explicit registration`, не превращая EvoHime в обязательный ML toolchain. Fine-tuning/LoRA/QLoRA и model merge допускаются как последующие adapter capabilities поверх того же pipeline, но не должны быть обязательны для первой реализации.

Функциональность должна сохранять local-first/self-contained характер: отсутствие конкретного adaptation backend возвращает typed `unavailable`, а не устанавливает Python/CUDA/внешние сервисы скрытно и не создаёт cloud-зависимость.

## Проблема

Текущий EvoHime уже имеет подходящие владельцы:

- `Local Model Runtime Manager` хранит `LocalModelDescriptor`, format/quantization, artifact hash/size, hardware fit и runtime identity;
- `Local Model Performance Calibration` связывает точный model artifact/runtime/hardware/config identity с измерениями;
- `Agent Benchmark Matrix` сравнивает versioned model/profile combinations и умеет reproducible deterministic/provider modes;
- hardware/resource pressure, durable background execution, provenance/evidence, approvals и authenticated IPC уже существуют.

Но между импортом исходной модели и регистрацией готового локального артефакта нет typed transformation lifecycle. EvoHime не умеет безопасно выразить: какой исходный model artifact преобразован, каким backend/toolchain, с какими параметрами, во что именно, какой hash получен, выдержал ли артефакт structural/runtime verification и как его качество/latency/memory изменились относительно baseline.

Без такого контура пользователь вынужден выполнять преобразование внешними скриптами вручную, после чего Core видит только финальный файл и теряет provenance, воспроизводимость и возможность fail-closed promotion.

## Архитектурная граница

```text
Local Model Runtime Manager
  -> Adaptation preflight
  -> frozen source/model/hardware/toolchain snapshot
  -> existing durable background execution owner
  -> allowlisted optional adapter process
  -> staged derived artifact
  -> structural/hash/runtime verification
  -> calibration + Agent Benchmark Matrix
  -> explicit promotion/registration into existing local model catalog
  -> typed IPC projection
```

### Ownership

- **Rust Core** владеет adaptation request, lifecycle, policy, resource admission, paths, hashes, provenance, cancellation/recovery, verification и promotion decision.
- **Local Model Runtime Manager** остаётся единственным владельцем managed local model descriptors/artifact identity. Не создавать второй model registry.
- **Agent Benchmark Matrix** остаётся единственным общим quality/evaluation owner. Не создавать adaptation-specific benchmark engine.
- **Local Model Performance Calibration** остаётся владельцем machine-specific performance evidence.
- **Durable background execution/resource-pressure** переиспользуются для долгих jobs; не добавлять второй scheduler/daemon.
- **Electron/React** отображает только bounded typed projection и отправляет explicit user actions.
- Реальная конвертация/квантизация/training выполняется только через optional versioned adapters. Python, CUDA, external runtimes и vendor toolchains не являются обязательной частью базового EvoHime package.

### Обязательный MVP

1. Import/resolve уже известного Core source artifact.
2. Deterministic preflight: format, size, hash, disk/RAM/VRAM, adapter availability, target compatibility.
3. Format conversion через allowlisted adapter, если она нужна целевому runtime.
4. Quantization как explicit operation с typed target profile.
5. Staging + SHA-256 + size/format verification.
6. Runtime smoke/probe через существующий local runtime boundary.
7. Calibration и benchmark derived model против frozen baseline.
8. Explicit promotion в существующий local model catalog только после gates.

### Optional follow-up capabilities на том же pipeline

- LoRA/QLoRA adapter training;
- merge trained adapter into base model;
- bounded model merge recipes;
- additional quantization backends/formats.

Эти операции не должны расширять authority pipeline и не требуют отдельного storage/runtime.

## Typed contracts

Названия ориентировочные и могут быть приведены к текущему module naming без изменения семантики.

```rust
enum AdaptationOperationV1 {
    Convert,
    Quantize,
    TrainAdapter,
    MergeAdapter,
    MergeModels,
}

enum AdaptationJobStateV1 {
    Created,
    Preflighted,
    WaitingForResources,
    Running,
    Verifying,
    Benchmarking,
    ReadyForPromotion,
    Promoted,
    Rejected,
    Cancelled,
    Failed,
    Interrupted,
}

struct LocalModelAdaptationRequestV1 {
    schema_version: u32,
    job_id: String,
    idempotency_key: String,
    source_model_id: String,
    source_model_revision: u64,
    source_artifact_hash: String,
    operation: AdaptationOperationV1,
    target: AdaptationTargetV1,
    adapter_ref: AdaptationAdapterRef,
    resource_policy: AdaptationResourcePolicyV1,
    benchmark_suite_ref: Option<String>,
    user_approval_ref: Option<String>,
}

struct AdaptationTargetV1 {
    format: String,
    quantization: Option<String>,
    runtime_id: Option<String>,
    context_limit: Option<u32>,
}

struct AdaptationAdapterRef {
    adapter_id: String,
    version: String,
    executable_hash: String,
    capability: AdaptationOperationV1,
}

struct DerivedModelEvidenceV1 {
    source_artifact_hash: String,
    output_artifact_hash: String,
    adapter_hash: String,
    canonical_parameters_hash: String,
    hardware_fingerprint: String,
    format_probe: VerificationStatus,
    runtime_probe: VerificationStatus,
    calibration_ref: Option<String>,
    benchmark_report_ref: Option<String>,
    regression_status: RegressionStatus,
}
```

Raw tool stdout/stderr, training data, prompts, secrets и arbitrary command lines не входят в renderer projection.

## Adapter contract

Каждый adaptation backend должен иметь immutable descriptor с:

- exact executable/script/package identity;
- version + SHA-256;
- supported operations;
- supported input/output formats;
- supported quantization schemes;
- CPU/GPU requirements;
- resumability semantics;
- expected output contract;
- network requirement (`none` по умолчанию);
- deterministic/non-deterministic marker;
- command schema с typed arguments.

Core формирует argv напрямую. Shell interpretation, arbitrary executable path, arbitrary flags и пользовательская строка команды запрещены.

Adapter capability отсутствует или несовместима -> typed `unavailable_adapter`, без silent fallback на другой toolchain.

## Lifecycle

1. **Created** — canonical request принят и idempotency закреплена.
2. **Preflighted** — source hash/revision, hardware snapshot, disk reservation, adapter identity и target contract проверены.
3. **WaitingForResources** — job может ждать bounded resource window через существующий pressure/admission owner.
4. **Running** — adapter выполняется в Core-owned staging boundary.
5. **Verifying** — output существует, bounded по размеру, формат определяется фактически, hash вычислен, provenance заморожен.
6. **Benchmarking** — derived artifact проходит runtime probe, calibration и frozen benchmark suite.
7. **ReadyForPromotion** — все обязательные gates пройдены, но active model/catalog ещё не изменён.
8. **Promoted** — отдельное explicit действие создаёт/обновляет canonical `LocalModelDescriptor` revision.

Ни один derived artifact не становится preferred/active моделью автоматически.

## Persistence

Не создавать отдельный storage engine.

SQLite хранит только bounded job/evidence metadata в существующем Core-owned database:

- job id/revision/state;
- source descriptor/model/artifact refs;
- exact adapter identity/hash/version;
- canonical parameter hash;
- target format/quantization;
- staging/final managed locator metadata;
- output hash/size;
- hardware snapshot hash;
- verification/calibration/benchmark refs;
- timestamps, cancellation/recovery disposition и reason codes.

Большие model weights не кладутся blob-ом в SQLite и не проходят через task `ArtifactStore`. Производный model artifact должен использовать тот же managed local-model artifact boundary/path policy, который обслуживает `Local Model Runtime Manager`; pipeline лишь добавляет staging/publish semantics и provenance.

Training dataset для optional fine-tuning остаётся отдельным referenced source с hash/provenance; сырой dataset не копируется в generic job metadata.

## Atomic publication

Финальный файл не считается managed model до выполнения всех шагов:

1. write в Core-owned staging path;
2. flush/close;
3. размер и hash;
4. format probe;
5. runtime compatibility probe;
6. atomic move/publish в managed model location;
7. transactional update descriptor/evidence metadata.

Crash между staging и publish не должен создавать half-registered модель.

## Concurrency и resource admission

Heavy transformation является resource-sensitive operation.

- Core резервирует disk budget до запуска.
- Одновременно допускается bounded число adaptation jobs.
- GPU-heavy jobs сериализуются/ограничиваются по существующему hardware/resource-pressure owner.
- CPU threads/RAM/VRAM/temp disk limits фиксируются в immutable run snapshot.
- Adapter process не может самостоятельно расширять resource budget.
- Cancellation распространяется до process boundary; child process tree должен завершаться через существующие Windows process/Job Object semantics.
- Hashing/large file validation выполняются streaming, без загрузки multi-GB model artifact целиком в RAM.

## Security invariants

- `SOURCE_HASH_BEFORE_TRANSFORM`: преобразование разрешено только для повторно проверенного source artifact/revision.
- `ADAPTER_IDENTITY_IS_FROZEN`: executable/toolchain hash и версия фиксируются до запуска и не могут подмениться в середине job.
- `NO_ARBITRARY_COMMAND`: никакого shell/raw command line из renderer/model/user input.
- `NETWORK_DENY_BY_DEFAULT`: adaptation backend не получает сеть без отдельного declared capability + policy approval; обычная quantization/conversion работает offline.
- `NO_SILENT_TOOLCHAIN_INSTALL`: Python/CUDA/package manager/toolchain не устанавливаются автоматически.
- `NO_AUTOMATIC_PROMOTION`: успешный transform/benchmark не меняет active/preferred model без explicit Core action.
- `NO_UNVERIFIED_DERIVED_MODEL`: derived artifact с unknown/corrupt/failed format/runtime verification не регистрируется как ready.
- `BENCHMARK_BASELINE_IS_FROZEN`: source и derived сравниваются на одной версии suite/policy/profile, иначе результат `incomparable`, а не pass.
- `TRAINING_DATA_DOES_NOT_ESCALATE_TRUST`: optional fine-tune output не получает `ManagedVerified` только потому, что процесс завершился успешно; trust требует artifact + provenance + verification gates.

### Untrusted

- импортированный model artifact до hash/descriptor validation;
- external adapter output;
- tool stdout/stderr/progress;
- claimed model format/quantization metadata;
- training datasets;
- adapter-discovered telemetry.

### Trusted после Core verification

- frozen source descriptor + artifact hash;
- verified adapter descriptor/hash;
- normalized transformation parameters;
- verified derived artifact hash/size/format;
- Core-owned benchmark/calibration refs;
- promoted `LocalModelDescriptor` revision.

Renderer metadata не является источником trust.

## Crash recovery

Startup reconciliation должен быть deterministic:

- `Created/Preflighted/WaitingForResources` можно безопасно requeue/cancel по existing background policy;
- `Running` после Core/process crash не возобновляется вслепую: если adapter не объявил versioned resumable checkpoint contract, job становится `Interrupted`;
- resumable operation может продолжаться только при совпадении source hash, adapter hash, canonical params и checkpoint hash;
- `Verifying/Benchmarking` можно повторить идемпотентно поверх уже полученного immutable artifact;
- staging output с неправильным hash/format удаляется/карантинится bounded cleanup policy;
- `ReadyForPromotion` переживает restart и не выполняет promotion автоматически;
- published artifact без committed descriptor reconciles по journal/transaction marker, а не угадывается по имени файла.

## Observability

Core публикует bounded projection:

```rust
struct LocalModelAdaptationProjectionV1 {
    job_id: String,
    state: AdaptationJobStateV1,
    operation: AdaptationOperationV1,
    source_model_id: String,
    target_format: String,
    target_quantization: Option<String>,
    progress: Option<u8>,
    resource_status: ResourceAdmissionProjection,
    verification_status: VerificationStatus,
    benchmark_status: Option<RegressionStatus>,
    reason_code: Option<String>,
}
```

Логи/IPC не должны содержать raw model bytes, dataset rows, secrets или unbounded adapter output. Progress может быть coalesced; terminal state/evidence обязаны быть durable.

## Integration с существующими подсистемами

- `local_model_runtime_manager.rs`: source/derived descriptors, managed artifact identity, fit и promotion;
- `local_model_performance_calibration.rs`: post-transform machine-specific performance evidence;
- `agent_benchmark_matrix.rs`: frozen baseline/derived quality comparison;
- hardware fit / resource pressure: preflight и runtime admission;
- durable background execution: long-running lifecycle/recovery;
- verification evidence ledger: format/runtime/benchmark evidence refs;
- approvals/execution policy: explicit start и promotion для potentially expensive/unsafe operations;
- authenticated desktop IPC: projection-only UI;
- existing update/module mechanism: optional adapter binaries поставляются как versioned modules только если решено их bundle-ить, без второго updater.

### Contract/unit

- canonical request/parameter hashing;
- legal/illegal state transitions;
- adapter descriptor validation;
- unsupported format/quantization;
- source hash mismatch;
- stale source revision;
- disk/resource admission;
- idempotency replay;
- frozen baseline comparability.

### Integration

- fake converter produces verified deterministic artifact;
- fake quantizer -> derived descriptor -> runtime probe -> benchmark -> explicit promotion;
- cancellation terminates child process tree;
- adapter executable hash changes after preflight -> fail closed;
- output oversize/corrupt/truncated -> no promotion;
- restart during Running -> `Interrupted` unless valid resumable checkpoint;
- restart during Verifying/Benchmarking -> safe idempotent continuation;
- `ReadyForPromotion` survives restart without automatic activation;
- concurrent heavy jobs respect resource bounds;
- low disk / RAM / VRAM prevents unsafe start;
- IPC redaction/size bounds.

### Windows acceptance

- paths with spaces/Unicode;
- locked source/output files;
- antivirus/file-lock transient behavior does not produce false success;
- long-path handling within existing Windows path policy;
- process cancellation/restart leaves no orphan child process;
- managed artifact atomic publish on NTFS.

## Критерии готовности

- [ ] Core имеет versioned adaptation contract и не создаёт параллельный model registry/runtime/storage.
- [ ] MVP поддерживает supervised conversion + quantization через хотя бы один optional Windows-capable adapter.
- [ ] Source, adapter, params и derived artifact имеют immutable hash provenance.
- [ ] Derived artifact проходит structural + runtime verification до catalog registration.
- [ ] Hardware fit/resource pressure участвуют в preflight/admission.
- [ ] Calibration и Agent Benchmark Matrix сравнивают derived model с frozen baseline; incomparable evidence не выдаётся за improvement.
- [ ] Promotion в `LocalModelRuntimeManager` только explicit и fail-closed.
- [ ] Crash/cancel/restart не создают duplicate transforms, half-registered models или orphan process trees.
- [ ] Renderer остаётся projection-only; raw command/toolchain authority отсутствует.
- [ ] Базовый EvoHime не требует Python/CUDA/cloud для работы.
- [ ] Focused unit/integration/Windows tests, canonical docs и release evidence обновлены.

## Non-goals

- собственный training framework или CUDA kernel stack;
- обязательная установка Python/Conda/CUDA;
- cloud training control plane;
- автоматическая загрузка произвольных моделей/datasets из сети;
- arbitrary shell hooks/scripts;
- второй model catalog, benchmark engine, artifact DB или updater;
- automatic model promotion/routing switch;
- попытка гарантировать, что меньшая quantization всегда приемлема по качеству;
- обязательный fine-tuning в первом MVP.
